fedimint-client/src/backup.rs

```diff
@@ -189,33 +262,56 @@ impl Client {
         })
     }
 
-    /// Create a backup, include provided `metadata`, and encrypt it with a
-    /// (derived) client root key
-    pub async fn create_encrypted_backup(
-        &self,
-        metadata: Metadata,
-    ) -> Result<EncryptedClientBackup> {
-        let plaintext = self.create_backup(metadata).await?;
-        plaintext.encrypt_to(&self.get_derived_backup_encryption_key())
+    async fn load_previous_backup(&self) -> Option<ClientBackup> {
+        let mut dbtx = self.db.begin_transaction_nc().await;
+        dbtx.ensure_global().expect("global tx");
+        dbtx.get_value(&LastBackupKey).await
+    }
+
+    async fn store_last_backup(&self, backup: &ClientBackup) {
+        let mut dbtx = self.db.begin_transaction().await;
+        dbtx.ensure_global().expect("global tx");
+        dbtx.insert_entry(&LastBackupKey, backup).await;
+        dbtx.commit_tx().await;
     }
 
     /// Prepare an encrypted backup and send it to federation for storing
     pub async fn backup_to_federation(&self, metadata: Metadata) -> Result<()> {
-        let backup = self.create_encrypted_backup(metadata).await?;
+        let last_backup = self.load_previous_backup().await;
+        let new_backup = self.create_backup(metadata).await?;
+
+        let new_backup = new_backup.validate_and_fallback_module_backups(last_backup.as_ref());
+
+        let encrypted = new_backup.encrypt_to(&self.get_derived_backup_encryption_key())?;
+
+        self.validate_backup(&encrypted)?;
 
-        self.upload_backup(backup).await?;
+        self.store_last_backup(&new_backup).await;
 
+        self.upload_backup(&encrypted).await?;
+
+        Ok(())
+    }
+
+    /// Validate backup before sending it to federation
+    pub fn validate_backup(&self, backup: &EncryptedClientBackup) -> Result<()> {
+        if BACKUP_REQUEST_MAX_PAYLOAD_SIZE_BYTES < backup.len() {
```

Comment:
I haven't reviewed all of this code so apologies if this has been answered before, but why do we have a max size limit for both the encrypted and plaintext representations of the backup?

Reply:
Total backup limit is for the encrypted version. Per-module for unencrypted.